src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -1497,54 +1497,43 @@ async ValueTask<int> InternalFillBufferAsync(TReadAdapter adap, ValueTask<int> t
             }
         }
 
-        private ValueTask WriteAsyncInternal<TWriteAdapter>(TWriteAdapter writeAdapter, ReadOnlyMemory<byte> buffer)
+        private async Task WriteAsyncInternal<TWriteAdapter>(TWriteAdapter writeAdapter, ReadOnlyMemory<byte> buffer)
             where TWriteAdapter : struct, ISslWriteAdapter
         {
             CheckThrow(authSuccessCheck: true, shutdownCheck: true);
 
             if (buffer.Length == 0 && !SslStreamPal.CanEncryptEmptyMessage)
             {
                 // If it's an empty message and the PAL doesn't support that, we're done.
-                return default;
+                return;
             }
 
             if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
                 throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(WriteAsync), "write"));
             }
 
-            ValueTask t = buffer.Length < MaxDataSize ?
+            try
+            {
+                ValueTask t = buffer.Length < MaxDataSize ?
                     WriteSingleChunk(writeAdapter, buffer) :
                     new ValueTask(WriteAsyncChunked(writeAdapter, buffer));
-
-            if (t.IsCompletedSuccessfully)
```

Comment:
Why was this optimization removed?

Reply:
Because it's not measurable given everything else going on, and as written it's incorrect: if WriteSingleChunk, for example, throws an exception synchronously, then we never get here and avoid resetting state as is necessary.  At which point we need to add back a try/catch, and the resetting logic, all of which is already on the existing fallback path.